File: reference/tokens.py
from enum import Enum, auto
from dataclasses import dataclass
from typing import Any


class TokenKind(Enum):
    # Literals
    INT = auto()
    FLOAT = auto()
    STRING = auto()
    CHAR = auto()
    BOOL = auto()
    IDENT = auto()
    UNIT = auto()

    # Keywords
    FN = auto()
    LET = auto()
    MUT = auto()
    STRUCT = auto()
    ENUM = auto()
    IMPL = auto()
    TYPE = auto()
    PUB = auto()
    IF = auto()
    ELSE = auto()
    WHILE = auto()
    FOR = auto()
    IN = auto()
    MATCH = auto()
    RETURN = auto()
    BREAK = auto()
    CONTINUE = auto()
    IMPORT = auto()
    DOMAIN = auto()
    SPAWN = auto()
    AWAIT = auto()
    CHAN = auto()
    TRUE = auto()
    FALSE = auto()
    SELF = auto()
    SUPER = auto()
    PANIC = auto()
    UNREACHABLE = auto()
    MAYBE = auto()
    SOME = auto()
    NONE = auto()
    OK = auto()
    ERR = auto()

    # Operators
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()
    PERCENT = auto()
    EQ = auto()
    NEQ = auto()
    LT = auto()
    GT = auto()
    LTE = auto()
    GTE = auto()
    AND = auto()
    OR = auto()
    NOT = auto()
    ASSIGN = auto()
    PLUS_ASSIGN = auto()
    MINUS_ASSIGN = auto()
    STAR_ASSIGN = auto()
    SLASH_ASSIGN = auto()
    PERCENT_ASSIGN = auto()

    # Delimiters
    LPAREN = auto()
    RPAREN = auto()
    LBRACE = auto()
    RBRACE = auto()
    LBRACKET = auto()
    RBRACKET = auto()
    SEMICOLON = auto()
    COLON = auto()
    DOUBLE_COLON = auto()
    COMMA = auto()
    DOT = auto()
    DOUBLE_DOT = auto()
    ARROW = auto()
    FAT_ARROW = auto()
    AMP = auto()
    AMP_MUT = auto()
    QUESTION = auto()
    PIPE = auto()
    UNDERSCORE = auto()
    BANG = auto()

    # Special
    EOF = auto()
    NEWLINE = auto()


KEYWORDS = {
    "fn": TokenKind.FN,
    "let": TokenKind.LET,
    "mut": TokenKind.MUT,
    "struct": TokenKind.STRUCT,
    "enum": TokenKind.ENUM,
    "impl": TokenKind.IMPL,
    "type": TokenKind.TYPE,
    "pub": TokenKind.PUB,
    "if": TokenKind.IF,
    "else": TokenKind.ELSE,
    "while": TokenKind.WHILE,
    "for": TokenKind.FOR,
    "in": TokenKind.IN,
    "match": TokenKind.MATCH,
    "return": TokenKind.RETURN,
    "break": TokenKind.BREAK,
    "continue": TokenKind.CONTINUE,
    "import": TokenKind.IMPORT,
    "domain": TokenKind.DOMAIN,
    "spawn": TokenKind.SPAWN,
    "await": TokenKind.AWAIT,
    "chan": TokenKind.CHAN,
    "true": TokenKind.TRUE,
    "false": TokenKind.FALSE,
    "self": TokenKind.SELF,
    "super": TokenKind.SUPER,
    "panic": TokenKind.PANIC,
    "unreachable": TokenKind.UNREACHABLE,
    "maybe": TokenKind.MAYBE,
    "Some": TokenKind.SOME,
    "None": TokenKind.NONE,
    "Ok": TokenKind.OK,
    "Err": TokenKind.ERR,
}


@dataclass
class Span:
    line: int
    col: int
    length: int = 1

    def end_col(self) -> int:
        return self.col + self.length


@dataclass
class Token:
    kind: TokenKind
    value: str
    span: Span

    def __repr__(self) -> str:
        if self.value:
            return f"Token({self.kind.name}, {self.value!r}, L{self.span.line}:C{self.span.col})"
        return f"Token({self.kind.name}, L{self.span.line}:C{self.span.col})"

File: compiler/src/tokens.rs
use std::fmt;

#[derive(Debug, Clone, PartialEq)]
pub enum TokenKind {
    // Literals
    Integer(i64),
    Float(f64),
    String(String),
    Bool(bool),
    
    // Identifiers
    Identifier(String),
    
    // Keywords
    Fn,
    Let,
    Mut,
    If,
    Else,
    While,
    Return,
    Print,
    True,
    False,
    Struct,
    Enum,
    Domain,
    Spawn,
    Await,
    Chan,
    New,
    Some,
    None,
    Maybe,
    Type,
    As,
    In,
    Loop,
    Break,
    Continue,
    
    // Operators
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Eq,
    Neq,
    Lt,
    Gt,
    Lte,
    Gte,
    And,
    Or,
    Not,
    Assign,
    
    // Delimiters
    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    Comma,
    Colon,
    Semicolon,
    Arrow,
    DoubleColon,
    Dot,
    Question,
    
    // Special
    Eof,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub kind: TokenKind,
    pub line: usize,
    pub column: usize,
}

impl Token {
    pub fn new(kind: TokenKind, line: usize, column: usize) -> Self {
        Token { kind, line, column }
    }
}

impl fmt::Display for TokenKind {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            TokenKind::Integer(n) => write!(f, "{}", n),
            TokenKind::Float(n) => write!(f, "{}", n),
            TokenKind::String(s) => write!(f, "\"{}\"", s),
            TokenKind::Bool(b) => write!(f, "{}", b),
            TokenKind::Identifier(s) => write!(f, "{}", s),
            TokenKind::Fn => write!(f, "fn"),
            TokenKind::Let => write!(f, "let"),
            TokenKind::Mut => write!(f, "mut"),
            TokenKind::If => write!(f, "if"),
            TokenKind::Else => write!(f, "else"),
            TokenKind::While => write!(f, "while"),
            TokenKind::Return => write!(f, "return"),
            TokenKind::Print => write!(f, "print"),
            TokenKind::True => write!(f, "true"),
            TokenKind::False => write!(f, "false"),
            TokenKind::Struct => write!(f, "struct"),
            TokenKind::Enum => write!(f, "enum"),
            TokenKind::Domain => write!(f, "domain"),
            TokenKind::Spawn => write!(f, "spawn"),
            TokenKind::Await => write!(f, "await"),
            TokenKind::Chan => write!(f, "chan"),
            TokenKind::New => write!(f, "new"),
            TokenKind::Some => write!(f, "Some"),
            TokenKind::None => write!(f, "None"),
            TokenKind::Maybe => write!(f, "maybe"),
            TokenKind::Type => write!(f, "type"),
            TokenKind::As => write!(f, "as"),
            TokenKind::In => write!(f, "in"),
            TokenKind::Loop => write!(f, "loop"),
            TokenKind::Break => write!(f, "break"),
            TokenKind::Continue => write!(f, "continue"),
            TokenKind::Plus => write!(f, "+"),
            TokenKind::Minus => write!(f, "-"),
            TokenKind::Star => write!(f, "*"),
            TokenKind::Slash => write!(f, "/"),
            TokenKind::Percent => write!(f, "%"),
            TokenKind::Eq => write!(f, "=="),
            TokenKind::Neq => write!(f, "!="),
            TokenKind::Lt => write!(f, "<"),
            TokenKind::Gt => write!(f, ">"),
            TokenKind::Lte => write!(f, "<="),
            TokenKind::Gte => write!(f, ">="),
            TokenKind::And => write!(f, "&&"),
            TokenKind::Or => write!(f, "||"),
            TokenKind::Not => write!(f, "!"),
            TokenKind::Assign => write!(f, "="),
            TokenKind::LParen => write!(f, "("),
            TokenKind::RParen => write!(f, ")"),
            TokenKind::LBrace => write!(f, "{{"),
            TokenKind::RBrace => write!(f, "}}"),
            TokenKind::LBracket => write!(f, "["),
            TokenKind::RBracket => write!(f, "]"),
            TokenKind::Comma => write!(f, ","),
            TokenKind::Colon => write!(f, ":"),
            TokenKind::Semicolon => write!(f, ";"),
            TokenKind::Arrow => write!(f, "->"),
            TokenKind::DoubleColon => write!(f, "::"),
            TokenKind::Dot => write!(f, "."),
            TokenKind::Question => write!(f, "?"),
            TokenKind::Eof => write!(f, "EOF"),
        }
    }
}
